[Shopify] Connector - Staff to Salesperson mapping

AndriusA-Companial opened the pull request:
This pull request does not have a related issue as it's part of the delivery for development agreed directly with [@AndreiPanko](https://github.com/AndreiPanko)

* New functionality is only available for B2B stores. Hide (don’t use if the store is not B2B, see example – Companies/ Catalogs)
* Create the new Shopify Staff Mapping page (similar to Languages/Markets).
* Add action to read staff list. Allow user to map Staff entry to the Salesperson/purchaser entry. 
* When importing orders, for B2B stores also import Staff node (id only). Only for B2B stores to avoid errors during calls placed by non-plus stores. 
* Add the Sales Person Code field to the Shopify Order header table/page and initialize during mapping routine (similar to Map Customer / Payment Method). 
* When converting order from Shopify Order to Business Central sales document populate the Sales Person in the created sales order, similar to Payment Method. 
* Out of scope – export sales person to Shopify as Staff

AndriusAndrulevicius committed:
Staff to Sales person mapping

New functionality is only available for B2B stores. Hide (don’t use if the store is not B2B, see example – Companies/ Catalogs)
Create the new Shopify Staff Mapping page (similar to Languages/Markets).
Add action to read staff list. Allow user to map Staff entry to the Salesperson/purchaser entry.
When importing orders, for B2B stores also import Staff node (id only). Only for B2B stores to avoid errors during calls placed by non-plus stores.
Add the Sales Person Code field to the Shopify Order header table/page and initialize during mapping routine (similar to Map Customer / Payment Method).
When converting order from Shopify Order to Business Central sales document populate the Sales Person in the created sales order, similar to Payment Method.
Out of scope – export sales person to Shopify as Staff

GediminasGaubys commented on Apps/W1/Shopify/app/app.json:
```diff
@@ -4,7 +4,7 @@
   "brief": "Seamless connection between Shopify and Dynamics 365 Business Central will synchronize order, stock, and customer information to ensure that merchants can fulfill orders faster and better serve their customers.",
   "description": "Connecting Dynamics 365 Business Central with Shopify will help merchants implement more agile online business processes, while keeping people focused on selling. With support for multitier pricing structures and multiple currencies, companies, and entities, Business Central will support multiple Shopify store scenarios with ease. By connecting Shopify and Business Central, you  ...
-  "version": "27.0.0.0",
+  "version": "27.0.33055.1",
```
Revert the version change

GediminasGaubys commented on Apps/W1/Shopify/app/src/Order handling/Codeunits/ShpfyImportOrder.Codeunit.al:
```diff
@@ -1,5 +1,6 @@
 namespace Microsoft.Integration.Shopify;
 
+using Microsoft.Integration.Shopify;
```
Is this really needed, since this is Microsoft.Integration.Shopify namespace

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Codeunits/ShpfyGQLGetStaffMembers.Codeunit.al:
```diff
@@ -0,0 +1,26 @@
+    /// <returns>The GraphQL query as a text string.</returns>
+    internal procedure GetGraphQL(): Text
+    begin
+        exit('{"query":"{ staffMembers (first: 1) { edges { node { accountType active email exists firstName id initials isShopOwner lastName locale name phone } } pageInfo { hasPreviousPage hasNextPage startCursor endCursor } } }"}');
```
Why not first: 20, and then if there is more, another 20. Because it does not seem logical to first get only 1 and then make another call with 20

AndriusA-Companial replied:
This is a leftover from the testing (should have been 20)

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Tables/ShpfyStaffMember.Table.al:
```diff
@@ -0,0 +1,123 @@
+        field(1; "Shop Code"; Code[20])
+        {
+            Caption = 'Shop Code';
+            DataClassification = CustomerContent;
```
No need to declare it on every field, since you declared it above.

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Pages/ShpfyStaffMapping.Page.al:
```diff
@@ -0,0 +1,97 @@
+                }
+                field("Salesperson/Purchaser Code"; Rec."Salesperson Code")
+                {
+                    ToolTip = 'Specifies the sales person or purchaser code.';
```
Move the tooltips from the page to the table. Here we can leave it clean:
`field("Salesperson/Purchaser Code"; Rec."Salesperson Code") {}`

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Pages/ShpfyStaffMapping.Page.al:
```diff
@@ -0,0 +1,97 @@
+/// <summary>
+/// Page Shpfy Staff Mapping (ID 30171).
+/// </summary>
+page 30171 "Shpfy Staff Mapping"
```
Since this functionality can only be used if it is a B2B shop, it seems like the user can still open it and use the action which does nothing because of the exit.

AndriusAndrulevicius committed:
missing codeunit added

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Codeunits/ShpfyStaffMemberAPI.Codeunit.al:
```diff
@@ -0,0 +1,166 @@
+        repeat
+            JResponse := CommunicationMgt.ExecuteGraphQL(GraphQLType, Parameters);
+            if JResponse.IsObject() then
+                if this.ExtractStaffMembers(JResponse.AsObject(), Cursor, ShpfyShop.Code, TempShpfyStaffMember) then begin
```
Is this condition needed, since ExtractStaffMembers will always return true based on the functionality if no error occurs?

AndriusA-Companial replied:
this does just that - stops import if something goes wrong. I used https://github.com/GediminasGaubys/ALAppExtensions/blob/4d16226a5290f2c6ead4a9983f90a187e11224f7/Apps/W1/Shopify/app/src/Catalogs/Codeunits/ShpfyCatalogAPI.Codeunit.al#L83 as an example:
```al
if ExtractShopifyCatalogs(ShopifyCompany, JResponse.AsObject(), Cursor) then begin
    if Parameters.ContainsKey('After') then
        Parameters.Set('After', Cursor)
    else
        Parameters.Add('After', Cursor);
    GraphQLType := "Shpfy GraphQL Type"::GetNextCatalogs;
end else
    break;
```

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Codeunits/ShpfyStaffMemberAPI.Codeunit.al:
```diff
@@ -0,0 +1,166 @@
+                        Parameters.Add('After', Cursor);
+                    GraphQLType := GraphQLType::GetNextStaffMembers;
+                end else
+                    break;
```
the else part won't happen

AndriusA-Companial replied:
it will, if something goes wrong in the procedure in the condition (i.e. schema changes and we no longer have a field)

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Codeunits/ShpfyStaffMemberAPI.Codeunit.al:
```diff
@@ -0,0 +1,166 @@
+                TempShpfyStaffMember.Name := CopyStr(JsonHelper.GetValueAsText(JStaffMemberInfo, 'name'), 1, MaxStrLen(TempShpfyStaffMember.Name));
+                TempShpfyStaffMember."Locale" := CopyStr(JsonHelper.GetValueAsText(JStaffMemberInfo, 'locale'), 1, MaxStrLen(TempShpfyStaffMember."Locale"));
+                TempShpfyStaffMember.Phone := CopyStr(JsonHelper.GetValueAsText(JStaffMemberInfo, 'phone'), 1, MaxStrLen(TempShpfyStaffMember.Phone));
+                TempShpfyStaffMember.Insert();
```
I know that the default value is false, but let's declare it - Insert(false);

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Codeunits/ShpfyStaffMemberAPI.Codeunit.al:
```diff
@@ -0,0 +1,166 @@
+                        Modified := true;
+                    end;
+                    if Modified then
+                        StaffMember.Modify();
```
same here Modify(false)

GediminasGaubys commented on Apps/W1/Shopify/app/src/Staff/Codeunits/ShpfyStaffMemberAPI.Codeunit.al:
```diff
@@ -0,0 +1,166 @@
+        if StaffMember.FindSet() then
+            repeat
+                if not ProcessedStaffMembers.Contains(StaffMember.Id) then
+                    StaffMember.Delete();
```
Same here Delete(false)

AndriusAndrulevicius committed:
Comments fixed

AndriusAndrulevicius committed:
Automated tests for Staff functionality added

GediminasGaubys commented on Apps/W1/Shopify/app/src/Base/Codeunits/ShpfyCommunicationMgt.Codeunit.al:
```diff
@@ -579,6 +579,14 @@ codeunit 30103 "Shpfy Communication Mgt."
+    internal procedure DisableEventMocking()
+    begin
+        this.IsTestInProgress := false;
+    end;
```
Is this used? procedure SetTestInProgress does the same

AndriusA-Companial replied:
Procedure removed

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+    Subtype = Test;
+    TestPermissions = Disabled;
+    TestHttpRequestPolicy = BlockOutboundRequests;
+    EventSubscriberInstance = Manual;
```
Is this needed? We no longer use bindsubcribtion

AndriusA-Companial replied:
removed

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+        LibraryAssert: Codeunit "Library Assert";
+    begin
+        this.Initialize();
+        StaffMember.DeleteAll();
```
add parameter false

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+        this.Initialize();
+        StaffMember.DeleteAll();
+
+        // [Given] Staff exists in Shopify and is available for import
```
move given part above the Initialize

AndriusA-Companial replied:
done

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+
+        this.CreateNewStaffMember(ShpfyShop.Code, StaffMember);
+
+        // [Given] Staff was deleted in Shopify
```
same here, also in other places

AndriusA-Companial replied:
done

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+        LibraryAssert: Codeunit "Library Assert";
+    begin
+        this.Initialize();
+        StaffMember.DeleteAll();
```
false, also in other places

AndriusA-Companial replied:
done

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+        InitializeTest.SetDisableEventMocking();
+        ShpfyShop := InitializeTest.CreateShop();
+        ShpfyShop."B2B Enabled" := true;
+        ShpfyShop.Modify();
```
false

AndriusA-Companial replied:
done

GediminasGaubys commented on Apps/W1/Shopify/test/Staff/ShpfyStaffTest.Codeunit.al:
```diff
@@ -0,0 +1,296 @@
+        ShpfyShop."B2B Enabled" := true;
+        ShpfyShop.Modify();
+
+        // Disable Event Mocking 
```
Is this needed?

AndriusAndrulevicius committed:
Comments resolved

AndriusAndrulevicius committed:
Comments resolved

AndriusAndrulevicius committed:
Add LookupPageId and DrillDownPageId to Shpfy Staff Member table

AndriusAndrulevicius committed:
- Query to get Order reverted to original
- Updated logic to get Orders to facilitate that only B2B stores to get Staff Member information (Customer API used as reference)
- Table relation removed from Salesperson on Order Header
- Permission set ordered by name

AndriusAndrulevicius committed:
Rename parameter TempShpfyStaffMember to TempStaffMember in ExtractStaffMembers procedure for consistency

AndriusAndrulevicius committed:
Add GraphQL support for retrieving Shopify staff members and update related procedures

AndriusAndrulevicius committed:
Update GraphQL query and authentication scopes; add default email and phone fields to customer JSON

Groenbech96 committed:
Syncing with version 27.0.37019.0 (#28920)

Fixes
[AB#560539](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/560539)

---------

Co-authored-by: Magnus Hartvig Grønbech <magnushar@microsoft.com>

aholstrup1 committed:
Merge branch 'main' into dev/aan/staffmembers
